File: src/main/java/org/example/kalkulationsprogramm/domain/BelegPosition.java
package org.example.kalkulationsprogramm.domain;

import jakarta.persistence.*;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.Setter;
import lombok.ToString;

import java.math.BigDecimal;
import java.time.LocalDateTime;

/**
 * Eine einzelne Position auf einem Beleg (KI-extrahiert aus dem Scan).
 *
 * Wird nur fuer Belege mit {@link BelegAufteilungsModus#TEILWEISE} angelegt.
 * Der Mitarbeiter setzt am Handy per Checkbox {@link #istFuerFirma} pro
 * Position; BelegSplitService rechnet daraus die Firma-Summen am Beleg neu.
 *
 * mwstSatz wird pro Position gespeichert, damit gemischte MwSt-Saetze auf
 * einem Bon (z.B. 7% Lebensmittel + 19% Werkzeug) korrekt aufgeteilt werden.
 */
@Getter
@Setter
@Entity
@Table(name = "beleg_position")
@EqualsAndHashCode(of = "id")
@ToString(of = {"id", "sortierung", "beschreibung", "istFuerFirma"})
public class BelegPosition {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "beleg_id", nullable = false)
    private Beleg beleg;

    /** Reihenfolge auf dem Original-Bon (KI uebernimmt die visuelle Sortierung). */
    @Column(nullable = false)
    private int sortierung;

    @Column(nullable = false, length = 500)
    private String beschreibung;

    @Column(precision = 15, scale = 3)
    private BigDecimal menge;

    @Column(length = 20)
    private String einheit;

    @Column(precision = 15, scale = 4)
    private BigDecimal einzelpreis;

    @Column(name = "betrag_netto", precision = 15, scale = 2)
    private BigDecimal betragNetto;

    @Column(name = "betrag_brutto", precision = 15, scale = 2)
    private BigDecimal betragBrutto;

    @Column(name = "mwst_satz", precision = 5, scale = 2)
    private BigDecimal mwstSatz;

    @Column(name = "ist_fuer_firma", nullable = false)
    private boolean istFuerFirma = false;

    @Column(name = "erstellt_am", nullable = false)
    private LocalDateTime erstelltAm;

    @PrePersist
    void onCreate() {
        if (erstelltAm == null) {
            erstelltAm = LocalDateTime.now();
        }
    }
}

File: src/main/kotlin/org/example/kalkulationsprogramm/domain/BelegPosition.kt
package org.example.kalkulationsprogramm.domain

import jakarta.persistence.*
import java.math.BigDecimal
import java.time.LocalDateTime

@Entity
@Table(name = "beleg_position")
open class BelegPosition {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    open var id: Long? = null

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "beleg_id", nullable = false)
    open var beleg: Beleg? = null

    @Column(nullable = false)
    open var sortierung: Int = 0

    @Column(nullable = false, length = 500)
    open var beschreibung: String? = null

    @Column(precision = 15, scale = 3)
    open var menge: BigDecimal? = null

    @Column(length = 20)
    open var einheit: String? = null

    @Column(precision = 15, scale = 4)
    open var einzelpreis: BigDecimal? = null

    @Column(name = "betrag_netto", precision = 15, scale = 2)
    open var betragNetto: BigDecimal? = null

    @Column(name = "betrag_brutto", precision = 15, scale = 2)
    open var betragBrutto: BigDecimal? = null

    @Column(name = "mwst_satz", precision = 5, scale = 2)
    open var mwstSatz: BigDecimal? = null

    @Column(name = "ist_fuer_firma", nullable = false)
    open var istFuerFirma: Boolean = false

    open fun isIstFuerFirma(): Boolean = istFuerFirma

    @Column(name = "erstellt_am", nullable = false)
    open var erstelltAm: LocalDateTime? = null

}
